File: ios/GameController.swift
import Foundation
import Combine

final class GameController: ObservableObject {
    @Published var steer: Double = 0
    @Published var throttle: Double = 0
    @Published var brake: Double = 0
    @Published var buttons: [Bool] = Array(repeating: false, count: kMaxButtons)
    @Published var status: ConnectionState = .setup
    @Published var txRate: Int = 0
    @Published var target: String = ""

    let settings: AppSettings
    private let motion = MotionManager()
    private let network = NetworkManager()
    private let haptics = HapticsEngine()
    private var orientationFlipped = false
    private var lastUISteerTime: CFTimeInterval = 0

    init(settings: AppSettings) {
        self.settings = settings
        motion.settings = settings

        motion.onSoftLock = { [weak self] in
            guard let self, self.settings.hapticsEnabled else { return }
            DispatchQueue.main.async {
                self.haptics.softLock()
            }
        }

        motion.onSteer = { [weak self] value in
            guard let self else { return }
            var v = value
            if self.settings.autoInvert && self.orientationFlipped { v = -v }
            self.network.updateSteer(v)
            let now = CFAbsoluteTimeGetCurrent()
            if now - self.lastUISteerTime >= 1.0 / 60.0 {
                self.lastUISteerTime = now
                DispatchQueue.main.async { self.steer = v }
            }
        }
        network.onState = { [weak self] state in
            guard let self else { return }
            DispatchQueue.main.async {
                self.status = state
                if !state.isConnected { self.txRate = 0 }
            }
        }
        network.onRate = { [weak self] rate in
            DispatchQueue.main.async { self?.txRate = rate }
        }
        network.onHaptic = { [weak self] cue in
            guard let self else { return }
            
            // If the Mac sends an updated dynamic grip limit, feed it to the motion manager
            if let newLimit = cue.limit {
                self.motion.currentDynamicGripLimit = newLimit
            }
            
            guard self.settings.telemetryHaptics else { return }
            DispatchQueue.main.async {
                switch cue.haptic {
                case "lockup":    self.haptics.lockup()
                case "wheelspin": self.haptics.wheelspin(cue.intensity ?? 0.6)
                case "kerb":      self.haptics.kerb()
                case "offtrack":  self.haptics.offtrack(cue.intensity ?? 0.7)
                case "softLock":  self.haptics.softLock()
                default: break
                }
            }
        }
    }

    func start() { motion.start(); haptics.prepare() }
    func stop()  { motion.stop() }
    func setLandscapeFlipped(_ flipped: Bool) { orientationFlipped = flipped }

    func connect() {
        target = "\(settings.host):\(settings.port)"
        network.sendRateHz = settings.sendRateHz
        network.connect(host: settings.host, port: UInt16(clamping: settings.port))
        if settings.calibrateOnLaunch { calibrate() }
    }

    func connect(to mac: DiscoveredMac) {
        target = mac.name
        network.sendRateHz = settings.sendRateHz
        network.connect(endpoint: mac.endpoint)
        if settings.calibrateOnLaunch { calibrate() }
    }

    var targetLabel: String { target.isEmpty ? "\(settings.host):\(settings.port)" : target }
    func disconnect() { network.disconnect(); txRate = 0 }
    func calibrate() { motion.calibrate() }

    func setThrottleBrake(throttle: Double, brake: Double) {
        self.throttle = throttle
        self.brake = brake
        network.throttleEase = settings.throttleEase
        let t = curve(throttle, settings.throttleGamma)
        let b = curve(brake, settings.brakeGamma)
        if settings.invertPedals {
            network.updateThrottleBrake(throttle: b, brake: t)
        } else {
            network.updateThrottleBrake(throttle: t, brake: b)
        }
    }

    private func curve(_ v: Double, _ gamma: Double) -> Double {
        guard gamma != 1.0, v > 0 else { return v }
        return pow(min(max(v, 0), 1), gamma)
    }

    func buttonDown(_ index: Int) {
        guard buttons.indices.contains(index) else { return }
        let mode = ButtonMode(rawValue: settings.buttonModes[index]) ?? .hold
        switch mode {
        case .hold:   setButton(index, true)
        case .toggle: setButton(index, !buttons[index])
        case .tap:
            setButton(index, true)
            DispatchQueue.main.asyncAfter(deadline: .now() + 0.10) { [weak self] in
                self?.setButton(index, false)
            }
        }
    }

    func buttonUp(_ index: Int) {
        guard buttons.indices.contains(index) else { return }
        let mode = ButtonMode(rawValue: settings.buttonModes[index]) ?? .hold
        if mode == .hold { setButton(index, false) }
    }

    private func setButton(_ index: Int, _ pressed: Bool) {
        guard buttons.indices.contains(index) else { return }
        buttons[index] = pressed
        network.updateButton(name: "btn\(index + 1)", pressed: pressed)
    }
}

File: android/app/src/main/java/com/gyrowheel/GameController.kt
package com.gyrowheel

import android.app.Application
import android.os.Handler
import android.os.Looper
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateListOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.AndroidViewModel
import kotlin.math.pow

/**
 * Ties motion + network + haptics + discovery + settings together and exposes
 * Compose-observable state to the UI. Android twin of the iOS GameController.
 */
class GameController(app: Application) : AndroidViewModel(app) {
    val settings = AppSettings(app)
    val discovery = Discovery(app)

    private val motion = MotionManager(app)
    private val network = NetworkManager()
    private val haptics = HapticsEngine(app)
    private val main = Handler(Looper.getMainLooper())

    var steer by mutableStateOf(0.0); private set
    var throttle by mutableStateOf(0.0); private set
    var brake by mutableStateOf(0.0); private set
    val buttons = mutableStateListOf<Boolean>().apply { repeat(kMaxButtons) { add(false) } }
    var status by mutableStateOf<ConnectionState>(ConnectionState.Setup); private set
    var txRate by mutableStateOf(0); private set
    var target by mutableStateOf(""); private set

    private var orientationFlipped = false
    private var lastUiSteerNs = 0L

    init {
        motion.settings = settings

        motion.onSoftLock = {
            if (settings.hapticsEnabled) main.post { haptics.softLock() }
        }

        motion.onBackTap = { handleBackTapShift() }

        motion.onSteer = { value ->
            var v = value
            if (settings.autoInvert && orientationFlipped) v = -v
            network.updateSteer(v)
            val now = System.nanoTime()
            if (now - lastUiSteerNs >= 16_000_000L) { // ~60 Hz UI cap
                lastUiSteerNs = now
                main.post { steer = v }
            }
        }

        network.onState = { state ->
            main.post {
                status = state
                if (!state.isConnected) txRate = 0
            }
        }
        network.onRate = { rate -> main.post { txRate = rate } }
        network.onHaptic = { cue ->
            cue.limit?.let { motion.currentDynamicGripLimit = it }
            if (settings.telemetryHaptics) main.post {
                when (cue.haptic) {
                    "lockup" -> haptics.lockup()
                    "wheelspin" -> haptics.wheelspin(cue.intensity ?: 0.6)
                    "kerb" -> haptics.kerb()
                    "offtrack" -> haptics.offtrack(cue.intensity ?: 0.7)
                    "softLock" -> haptics.softLock()
                }
            }
        }
    }

    fun start() = motion.start()
    fun stop() = motion.stop()
    fun setLandscapeFlipped(flipped: Boolean) { orientationFlipped = flipped }

    val targetLabel: String get() = if (target.isEmpty()) "${settings.host}:${settings.port}" else target

    fun connect() {
        target = "${settings.host}:${settings.port}"
        network.sendRateHz = settings.sendRateHz
        network.connect(settings.host, settings.port)
        if (settings.calibrateOnLaunch) calibrate()
    }

    fun connect(mac: DiscoveredMac) {
        target = mac.name
        network.sendRateHz = settings.sendRateHz
        network.connect(mac.host, mac.port)
        if (settings.calibrateOnLaunch) calibrate()
    }

    fun disconnect() {
        network.disconnect()
        main.post { txRate = 0; status = ConnectionState.Setup }
    }

    fun calibrate() = motion.calibrate()

    fun setThrottleBrake(throttle: Double, brake: Double) {
        this.throttle = throttle
        this.brake = brake
        network.throttleEase = settings.throttleEase
        val t = curve(throttle, settings.throttleGamma)
        val b = curve(brake, settings.brakeGamma)
        if (settings.invertPedals) network.updateThrottleBrake(b, t)
        else network.updateThrottleBrake(t, b)
    }

    private fun curve(v: Double, gamma: Double): Double {
        if (gamma == 1.0 || v <= 0) return v
        return v.coerceIn(0.0, 1.0).pow(gamma)
    }

    fun buttonDown(index: Int) {
        if (index !in buttons.indices) return
        motion.suppressBackTap()
        when (ButtonMode.from(settings.buttonModes.getOrElse(index) { 0 })) {
            ButtonMode.HOLD -> setButton(index, true)
            ButtonMode.TOGGLE -> setButton(index, !buttons[index])
            ButtonMode.TAP -> {
                setButton(index, true)
                main.postDelayed({ setButton(index, false) }, 100)
            }
        }
    }

    fun buttonUp(index: Int) {
        if (index !in buttons.indices) return
        if (ButtonMode.from(settings.buttonModes.getOrElse(index) { 0 }) == ButtonMode.HOLD) setButton(index, false)
    }

    private fun setButton(index: Int, pressed: Boolean) {
        if (index !in buttons.indices) return
        buttons[index] = pressed
        network.updateButton("btn${index + 1}", pressed)
    }

    // MARK: - Back-tap paddle shifter
    private fun handleBackTapShift() {
        if (!settings.backTapShiftEnabled) return
        main.post {
            val up = throttle >= 0.5
            val index = if (up) settings.upshiftButton else settings.downshiftButton
            pulseShift(index)
            if (settings.shiftHaptics && settings.hapticsEnabled) haptics.shift(up)
        }
    }

    private fun pulseShift(index: Int) {
        if (index !in buttons.indices) return
        setButton(index, true)
        main.postDelayed({ setButton(index, false) }, 90)
    }

    override fun onCleared() {
        super.onCleared()
        network.disconnect()
        motion.stop()
        discovery.stop()
    }
}
